### PR TITLE
Removes need to maintain a Redis Set for Primary Keys

### PR DESCRIPTION
**Pull request by bsbodden**
Spring Data Redis uses a Redis Set to keep primary keys for a mapped entity and perform operations related to primary keys. The PR removes the PK set and rewrites all dependent operations using Search Queries or Aggregations. A side-effect of this refactoring is that now **ALL** mapped entities with `@RedisHash` or `@Document` will have an associated index.

### CODE REVIEW

File: redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java [hunk 1]
```diff
@@ -236,6 +239,39 @@ public <T> T get(Object id, String keyspace, Class<T> type) {
...
+
+      byte[] keyToDelete = createKey(asString(keyspace), asString(id));
+
+      redisOperations.execute((RedisCallback<Void>) connection -> {
+
+        connection.del(keyToDelete);
+
+        if (keepShadowCopy()) {
+          RedisPersistentEntity<?> persistentEntity = converter.getMappingContext().getPersistentEntity(type);
+          if (persistentEntity != null && persistentEntity.isExpiring()) {
```
Comment (slorello89) [inline redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java:263]: is there a bit of context here I'm missing? The boolean for `persistentEntity.isExpiring()` seems inverted. Unless the intention is to persist the ghosts of expired objects as they die?

File: redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java [hunk 2]
```diff
@@ -245,13 +281,37 @@ public <T> T get(Object id, String keyspace, Class<T> type) {
    */
   @Override
   public void deleteAllOf(String keyspace) {
+    Class<?> type = keyspaceToIndexMap.getEntityClassForKeyspace(keyspace);
```
Comment (slorello89) [inline redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java:284]: The performance implications of this are going to be brutal, I wonder if it wouldn't just make more sense to script a:

```
FT.DROPINDEX indexName DD
FT.CREATE indexName all the args
```

that way you're dropping down to a single round trip and ensuring that you are in fact killing that entire keyspace without any fear or races. Not a great solution but will probably be more performant & robust.

File: redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java [hunk 3]
```diff
@@ -261,32 +321,24 @@ public void deleteAllOf(String keyspace) {
...
-    }
-
-    offset = Math.max(0, offset);
-    if (rows > 0) {
-      keys = keys.subList((int) offset, Math.min((int) offset + rows, keys.size()));
-    }
-
-    for (byte[] key : keys) {
-      result.add(get(key, keyspace, type));
+    List<String> keys = getAllIds(keyspace, type);
```
Comment (slorello89) [inline redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java:330]: It looks like there's an extra step here, It seems to me you're running:

```
FT.SEARCH index * RETURN 1 id LIMIT offset rows
```

and then a sequence of:

```
HGETALL id
```

You could just short circuit the whole thing and just use:

```
FT.SEARCH index * LIMIT offset rows
```

File: redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java [hunk 4]
```diff
@@ -359,6 +411,45 @@ public void update(PartialUpdate<?> update) {
...
   }
 
+  /*
+   * (non-Javadoc)
+   * 
+   * @see org.springframework.data.keyvalue.core.KeyValueAdapter#count(java.lang.
+   * String)
+   */
+  @Override
+  public long count(String keyspace) {
```
Comment (slorello89) [inline redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java:421]: `FT.SEARCH index * LIMIT 0 0` works better for this sort of thing I think as it just pulls directly from the index metadata.

**Review by slorello89 — CHANGES_REQUESTED** [review]
Some suggestions as to how to make these operations more performant and straight forward - it's entirely possible I'm missing some context on why the hash operations are being enumerated the way they are?

File: redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java [hunk 5]
```diff
@@ -225,19 +225,17 @@ public <T> T delete(Object id, String keyspace, Class<T> type) {
...
-    if (keys.size() > 0) {
-      redisOperations.execute((RedisCallback<Void>) connection -> {
-        connection.del(keys.toArray(new byte[0][]));
-        return null;
-      });
+    Class<?> type = indexer.getEntityClassForKeyspace(keyspace);
+    Optional<String> maybeSearchIndex = indexer.getIndexName(keyspace);
+    if (maybeSearchIndex.isPresent()) {
+      SearchOperations<String> searchOps = modulesOperations.opsForSearch(maybeSearchIndex.get());
+      searchOps.dropIndex();
```
Comment (slorello89) [inline redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java:227]: Is this using the `FT.DROP`? or `FT.DROPINDEX DD` - I'd use the non-deprecated command.

**Review by slorello89 — CHANGES_REQUESTED** [review]

File: redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java (hunk 5 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/RedisEnhancedKeyValueAdapter.java:227]: It's mapped to `DROP("FT.DROP"),` in the Jedis version I'm stuck with there is no `FT.DROPINDEX` afaik

**Review by slorello89 — APPROVED** [review]
LGTM 👍

**Pull request merged**